planning/MartenV4.md

```diff
@@ -0,0 +1,405 @@
+
+        // There's a little bit of optimization for inline
+        // projections by doing this
+
+        // Other folks want different mechanics
+        public StreamState State { get; set; }
+
+        public IList<TEvent> Events { get; } = new List<TEvent>();
+
+
+        // Data in here while it's being appended, but moves up above when
+        // it's pushed into the database
+        public IList<TEventBase> Pending { get; } = new List<TEventBase>();
+    }
+
+    public interface IEvent<TEventBase>
+    {
+        Guid Id { get; set; }
+        int Version { get; set; }
+        long Sequence { get; set; }
+        TEventBase Data { get; set; }
+        DateTimeOffset Timestamp { get; set; }
+
+        // Additional, user-defined metadata properties
+        // like user id or correlation id or whatever
+
+    }
+
+    // This might be too heavyweight
+    public interface IMultitenantEvent<TEventBase>: IEvent<TEventBase>
+    {
+        string TenantId { get; set; }
+    }
+
+The custom stream and event types could be used to define segregations between different types of event logs within a single application as part
+of the partitioning/sharding/segmenting strategy
+
+
+
+The schema generation would conform to the shape of the base event and stream wrapper much the way that document storage does in the Doc Db features.
+
+### Table/Function Generation and Stream Identity
+
+TODO -- allow more flexibility in the identity of streams, so `int`, `long`, `Guid`, `string` just like the document types. Do more code generation
+to accomodate this.
+
+TODO -- there will be more columns & even new tables to track and coordinate the asynchronous projections
+
+### Event Store Partitioning or Sharding
+
+The event sourcing in Marten has limited scalability because of its dependence today upon appending events to 
+one table. Closely related to the storage is ways to segment the asynchronous projection support to parallelize
+the work across multiple threads or active nodes.
+
+Here are some ideas:
+
+* Allow the user to define multiple event logs in the database. This way you'd have multiple stream/event tables for completely
+  different kinds of streams
+* Postgresql native partitioning by tenant id where that's useful
+* Aggregation keys -- this is tightly coupled to the projections. If there's a projection that is done by something other than tenant id or stream id, try to persist the 
+  key for however the events/streams are grouped to stream and/or event storage. Thinking about data elements like "Region" or "Country" where you might be doing
+  aggregations across streams or tenants. Maybe that's useful?
+* Archive streams that reach some kind of termination event? Move everything to an archive table off to the side? Keeps the event log table cleaner
+* Artificial segmenting of the streams. Use some kind of algorithm that could see stream ids, and assign them to 2 or more "cohorts", then persist that "cohort key" to 
```

Comment:
That part I'm not getting ;)

Reply:
Let's say you have a stream type that you want to aggregate by stream, and it happens to be using `long` for its id type. Then say you only want two segments, "even" and "odd". You could do the per-stream aggregates in two running projection builders, one doing "even" streams and one handling "odd" streams to get more parallelization. That's the kind of artificial segmenting I mean here.

Reply:
I'm sorry, but I'm still not getting the idea. Maybe some more real world sample/business case would help?

Reply:
public class QuestPartyTransformer
{
    public QuestParty Create(QuestStarted started)
    {
        // build stuff
     }


     // sync mutation signature
     public void Apply(MonsterSlayed @event, QuestParty aggregate)
     {

     }

     // immutable signature
     public QuestParty Apply(MonsterSlayed @event, QuestParty aggregate)
     {

     }

      // and variable signatures for doing things async and with IoC services
}

Reply:
I'm looking to create artificial partitioning of streams just to help parallelize more of the async projections. Let's say you're using `int` for stream ids. In that case we could have one active projection builder working with "even numbered stream ids" and another projection builder in parallel working on "odd numbered stream ids"

That make sense?

Reply:
For this sample how you'd define which event field would be used as the Id to select the document?

About the artificial partitioning, I'm not sure in which case it would help to parallelize the processing (besides the cases in which the event order doesn't matter).

Reply:
You'd be splitting based on **stream**, not by event. So you'd still be able to process events within a stream by the stream version number.